File: app/src/main/java/com/denchic45/kts/utils/SearchKeysGenerator.java
package com.denchic45.kts.utils;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class SearchKeysGenerator {

    private final List<String> searchKeys = new ArrayList<>();

    public List<String> generateKeys(@NotNull String text) {
        return createListOfKeys(text);
    }

    public List<String> generateKeys(@NotNull String text, Filter filter) {
        return createListOfKeys(text).stream().filter(filter).collect(Collectors.toList());
    }

    @NotNull
    private List<String> createListOfKeys(@NotNull String text) {
        searchKeys.clear();
        text = text.toLowerCase();
        List<List<String>> wordKeysLists = new ArrayList<>();
        for (String word : text.split("\\s+")) {
            wordKeysLists.add(generateKeysByWord(word));
        }
        permutationWordKeyLists(wordKeysLists, 0);
        return searchKeys;
    }

    private void permutationWordKeyLists(@NotNull List<List<String>> wordKeysLists, int pos) {
        if (pos == wordKeysLists.size() - 1) {
            printKeys(wordKeysLists);
            return;
        }
        for (int i = pos; i < wordKeysLists.size(); i++) {
            Collections.swap(wordKeysLists, i, pos);
            permutationWordKeyLists(wordKeysLists, pos + 1);
            Collections.swap(wordKeysLists, i, pos);
        }
    }

    private void printKeys(@NotNull List<List<String>> wordKeysLists) {
        StringBuilder appendedWords = new StringBuilder();
        for (List<String> wordKeys : wordKeysLists) {
            searchKeys.addAll(wordKeys.stream().map(key -> appendedWords.toString() + key).collect(Collectors.toList()));
            appendedWords.append(wordKeys.get(wordKeys.size() - 1));
        }
    }

    @NotNull
    private List<String> generateKeysByWord(@NotNull String word) {
        List<String> wordKeys = new ArrayList<>();
        StringBuilder str = new StringBuilder();
        for (char c : word.toCharArray()) {
            str.append(c);
            wordKeys.add(str.toString());
        }
        return wordKeys;
    }

    @NotNull
    public static String formatInput(@NotNull String input) {
        return input.toLowerCase().replaceAll("\\s+", "");
    }

    public interface Filter extends Predicate<String> {
        @Override
        boolean test(String predicate);
    }

}

File: app/src/main/java/com/denchic45/kts/utils/SearchKeysGenerator.kt
package com.denchic45.kts.utils

import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors

class SearchKeysGenerator {
    private val searchKeys: MutableList<String> = ArrayList()
    fun generateKeys(text: String): List<String> {
        return createListOfKeys(text)
    }

    fun generateKeys(text: String, filter: (String) -> Boolean): List<String> {
        return createListOfKeys(text).stream().filter(filter).collect(Collectors.toList())
    }

    private fun createListOfKeys(text: String): List<String> {
        var text = text
        searchKeys.clear()
        text = text.lowercase(Locale.getDefault())
        val wordKeysLists: MutableList<List<String>?> = ArrayList()
        for (word in text.split("\\s+").toTypedArray()) {
            wordKeysLists.add(generateKeysByWord(word))
        }
        permutationWordKeyLists(wordKeysLists, 0)
        return searchKeys
    }

    private fun permutationWordKeyLists(wordKeysLists: List<List<String>?>, pos: Int) {
        if (pos == wordKeysLists.size - 1) {
            printKeys(wordKeysLists)
            return
        }
        for (i in pos until wordKeysLists.size) {
            Collections.swap(wordKeysLists, i, pos)
            permutationWordKeyLists(wordKeysLists, pos + 1)
            Collections.swap(wordKeysLists, i, pos)
        }
    }

    private fun printKeys(wordKeysLists: List<List<String>?>) {
        val appendedWords = StringBuilder()
        for (wordKeys in wordKeysLists) {
            searchKeys.addAll(
                wordKeys!!.stream().map { key: String -> appendedWords.toString() + key }
                    .collect(Collectors.toList()))
            appendedWords.append(wordKeys[wordKeys.size - 1])
        }
    }

    private fun generateKeysByWord(word: String): List<String> {
        val wordKeys: MutableList<String> = ArrayList()
        val str = StringBuilder()
        for (c in word.toCharArray()) {
            str.append(c)
            wordKeys.add(str.toString())
        }
        return wordKeys
    }

    interface Filter : Predicate<String?> {
        override fun test(predicate: String?): Boolean
    }

    companion object {
        fun formatInput(input: String): String {
            return input.lowercase(Locale.getDefault()).replace("\\s+".toRegex(), "")
        }
    }
}